analyzers/src/SonarAnalyzer.Common/Helpers/WildcardPatternMatcher.cs

```diff
@@ -0,0 +1,134 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+using System.IO;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Helpers;
+
+internal static class WildcardPatternMatcher
+{
+    public static bool IsMatch(string pattern, string input) =>
+        !(string.IsNullOrWhiteSpace(pattern) || string.IsNullOrWhiteSpace(input))
+        && WildcardPattern.Create(pattern).Match(input);
+
+    /// <summary>
+    /// Copied from https://github.com/SonarSource/sonar-plugin-api/blob/a9bd7ff48f0f77811ed909070030678c443c975a/sonar-plugin-api/src/main/java/org/sonar/api/utils/WildcardPattern.java.
+    /// </summary>
+    private sealed class WildcardPattern
+    {
+        private const string SpecialChars = "()[]^$.{}+|";
+        private static readonly ConcurrentDictionary<string, WildcardPattern> Cache = new();
+        private readonly Regex pattern;
+
+        private WildcardPattern(string pattern, string directorySeparator) =>
+            this.pattern = new Regex(ToRegexp(pattern, directorySeparator), RegexOptions.Compiled, RegexConstants.DefaultTimeout);
```

Comment:
`Compiled` is extremely expensive (around magninute of x100000 if I remember correctly, because it is emitting dynamic in-memory assembly). 
Don't use it, unless you've verified it is the only solution to a major problem we had. 
These patterns should be rather simple.
```suggestion
            this.pattern = new Regex(ToRegexp(pattern, directorySeparator), RegexOptions.None, RegexConstants.DefaultTimeout);
```

Reply:
Fixed by #6919

Reply:
That sounds roughly correct.

There's also the complexity of the matching in play a bit. The more complex (more slow) the regex is, and the bigger the data, the better. Here, we have small regexes on small number of files typically.